File: backend-node/src/products/repositories/product.repository.ts
import { Injectable } from '@nestjs/common';
import { PgService } from '../../database/pg.service';

export interface Product {
  id: number;
  name: string;
  description: string;
  price: string;
  image_url: string | null;
  category: string;
  stock: number;
  created_at: Date;
  updated_at: Date;
}

@Injectable()
export class ProductRepository {
  constructor(private readonly pgService: PgService) {}

  async findAll(): Promise<Product[]> {
    const result = await this.pgService.query(
      'SELECT * FROM products ORDER BY created_at DESC'
    );
    console.log(`[ProductRepository] Retrieved ${result.rows.length} products`);
    return result.rows;
  }

  async findById(id: number): Promise<Product | null> {
    const result = await this.pgService.query(
      'SELECT * FROM products WHERE id = $1',
      [id]
    );
    const product = result.rows[0] || null;
    console.log(`[ProductRepository] ${product ? 'Found' : 'Did not find'} product with id ${id}`);
    return product;
  }

  async findByIds(ids: number[]): Promise<Product[]> {
    const result = await this.pgService.query(
      'SELECT * FROM products WHERE id = ANY($1::int[])',
      [ids]
    );
    console.log(`[ProductRepository] Retrieved ${result.rows.length} of ${ids.length} products`);
    return result.rows;
  }

  async searchWithFilters(search?: string, category?: string): Promise<Product[]> {
    let query = 'SELECT * FROM products WHERE 1=1';
    const params: any[] = [];
    let paramIndex = 1;

    if (search) {
      params.push(`%${search}%`);
      query += ` AND (name ILIKE $${paramIndex} OR description ILIKE $${paramIndex})`;
      paramIndex++;
    }

    if (category) {
      params.push(category);
      query += ` AND category = $${paramIndex}`;
      paramIndex++;
    }

    query += ' ORDER BY created_at DESC';

    const result = await this.pgService.query(query, params);
    console.log(`[ProductRepository] Search returned ${result.rows.length} products (search: "${search}", category: "${category}")`);
    return result.rows;
  }

  async getCategories(): Promise<string[]> {
    const result = await this.pgService.query(
      'SELECT DISTINCT category FROM products WHERE category IS NOT NULL ORDER BY category'
    );
    const categories = result.rows.map(row => row.category).filter(Boolean);
    console.log(`[ProductRepository] Retrieved ${categories.length} distinct categories`);
    return categories;
  }
}

File: backend/app/repositories/product_repository.py
from typing import List, Optional
from app.database.connection import DatabaseConnection
from app.models.product import Product


class ProductRepository:
    def __init__(self, db: DatabaseConnection):
        self.db = db

    def find_all(self) -> List[dict]:
        """Get all products."""
        return self.db.query("SELECT * FROM products ORDER BY id")

    def find_by_id(self, product_id: int) -> Optional[dict]:
        """Get a product by ID."""
        return self.db.query_one(
            "SELECT * FROM products WHERE id = %s",
            (product_id,)
        )

    def find_by_ids(self, product_ids: List[int]) -> List[dict]:
        """Get products by list of IDs."""
        if not product_ids:
            return []
        placeholders = ','.join(['%s'] * len(product_ids))
        return self.db.query(
            f"SELECT * FROM products WHERE id IN ({placeholders})",
            tuple(product_ids)
        )

    def search_with_filters(
        self,
        search: Optional[str] = None,
        category: Optional[str] = None
    ) -> List[dict]:
        """Search products with optional filters."""
        conditions = []
        params = []

        if search:
            conditions.append("(name ILIKE %s OR description ILIKE %s)")
            search_pattern = f"%{search}%"
            params.extend([search_pattern, search_pattern])

        if category:
            conditions.append("category = %s")
            params.append(category)

        where_clause = " AND ".join(conditions) if conditions else "1=1"
        query = f"SELECT * FROM products WHERE {where_clause} ORDER BY id"

        return self.db.query(query, tuple(params) if params else None)

    def get_categories(self) -> List[str]:
        """Get distinct product categories."""
        results = self.db.query(
            "SELECT DISTINCT category FROM products WHERE category IS NOT NULL ORDER BY category"
        )
        return [r['category'] for r in results]
